30-indoor-air-quality-and-gas-leak/firmware/sparrow-application/aqi/aqi.cpp

```diff
@@ -0,0 +1,496 @@
+    MX_I2C2_Init();
+    result = (MY_I2C2_Ping(BME280_I2C_ADDR_PRIM, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT)
+           || MY_I2C2_Ping(BME280_I2C_ADDR_SEC, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT));
+    MX_I2C2_DeInit();
+
+    return result;
+}
+
+// Scheduled App One-Time Init
+bool aqiInit()
+{
+    // Register the app
+    schedAppConfig config = {
+        .name = "aqi",
+        .activationPeriodSecs = 60 * 60,    // 1 hour
+        .pollPeriodSecs = 15,
+        .activateFn = NULL,
+        .interruptFn = NULL,
+        .pollFn = aqiPoll,
+        .responseFn = aqiResponse,
+        .appContext = &aqiSensor,
+    };
+    if (schedRegisterApp(&config) < 0) {
+        return false;
+    }
+
+    // Success
+    return true;
+
+}
+
+bool isAlert(AQISensorReading& reading) {
+    return (ALERT_AQI_LEVEL && reading.aqi >= ALERT_AQI_LEVEL)
+    || (ALERT_ECO2_LEVEL && reading.eco2 >= ALERT_ECO2_LEVEL)
+    || (ALERT_TVOC_LEVEL && reading.tvoc >= ALERT_TVOC_LEVEL);
+}
+
+bool takeReading(AQISensorReading& reading) {
+    if (!aqiSensor.read(reading)) {
+        return false;
+    }
+
+    reading.alert = isAlert(reading);
+    return true;
+}
+
+bool shouldSendMonitoringEvent() {
+    // this could be used to send monitoring events less frequently than how often the app takes a sensor reading.
+    return !ALERTS_ONLY;
+}
+
+
+// Poller
+void aqiPoll(int appID, int state, void *appContext)
+{
+    // Unused parameter(s)
+    (void)appContext;
+
+    AQISensorReading sensorReading;
+
+    // Switch based upon state
+    switch (state) {
+
+    case STATE_ACTIVATED:
+        if (!templateRegistered) {
+            registerNotefileTemplate();
+            schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_ACTIVATED);
+            APP_PRINTF("aqi: template registration request\r\n");
+            break;
+        }
+        // fallthrough to monitoring immediately
+
+    case STATE_AQI_MONITORING:
+         if (takeReading(sensorReading)) {
+            if (sensorReading.isAlert()) {
+                addNote(sensorReading, true, ALERT_FIRST);
+                schedSetCompletionState(appID, STATE_AQI_ALERTING, STATE_AQI_MONITORING);
+            }
+            else if (shouldSendMonitoringEvent()) {
+                addNote(sensorReading, false, ALERT_NONE);
+                schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_AQI_MONITORING);
+            }
+        }
+        break;
+
+    case STATE_AQI_ALERTING:
+        if (takeReading(sensorReading)) {
+            if (sensorReading.isAlert()) {
+                if (shouldSendMonitoringEvent()) {
+                    addNote(sensorReading, false, ALERT_ONGOING);
+                    schedSetCompletionState(appID, STATE_AQI_ALERTING, STATE_AQI_ALERTING);
+                }
+            }
+            else {
+                addNote(sensorReading, true, ALERT_STANDDOWN);
+                schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_AQI_ALERTING);
```

Comment:
Are we stuck in this state? I don't see how it transitions out of `STATE_AQI_ALERTING`.

Reply:
The success state is `STATE_AQI_MONITORING` so once the final alert:3 is sent, it goes back to monitoring. If the message is not sent, it essentially retries until successful.

Reply:
I just misread the function call. All good here.